### PR TITLE
Accelerate stable-segmented-sort with CUB segmented sort

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Adds calls to `cub::DeviceSegmentedSort::StableSortPairs` for `cudf::segmented_sorted_order`.
Refactors libcudf segmented-sort and stable-segmented-sort functions into separate source files and moves the common code into `segmented_sort_impl.cuh`. 
Also adds gtests for `cudf::stable_segmented_sort`.

Reference https://github.com/rapidsai/cudf/issues/11729

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/sort/segmented_sort.cu [hunk 1]
```diff
@@ -28,165 +27,9 @@
 #include <thrust/binary_search.h>
 #include <thrust/iterator/counting_iterator.h>
 
-#include <cub/device/device_segmented_sort.cuh>
-
 namespace cudf {
 namespace detail {
 
-namespace {
```
Comment (davidwendt) [inline cpp/src/sort/segmented_sort.cu:36]: This code moved to `segmented_sort_impl.cuh`

File: cpp/src/sort/segmented_sort_impl.cuh [hunk 2]
```diff
@@ -0,0 +1,324 @@
+/*
```
Comment (davidwendt) [inline cpp/src/sort/segmented_sort_impl.cuh:1]: This code was moved from `segmented_sort.cu`

**Review by ttnghia — APPROVED** [review]

**Review by robertmaynard — APPROVED** [review]
CMake changes LGTM

File: cpp/src/sort/segmented_sort_impl.cuh [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+                 bool ascending,
+                 rmm::cuda_stream_view stream)
+  {
+    // CUB's segmented sort functions cannot accept iterators.
+    // We create a temporary column here for it to use.
+    auto temp_col =
+      cudf::detail::allocate_like(input, input.size(), mask_allocation_policy::NEVER, stream);
+    mutable_column_view output_view = temp_col->mutable_view();
+
+    // DeviceSegmentedSort is faster then DeviceSegmentedRadixSort at this time
```
Comment (PointKernel) [inline cpp/src/sort/segmented_sort_impl.cuh:79]:
```suggestion
    // DeviceSegmentedSort is faster than DeviceSegmentedRadixSort at this time
```

**Review by PointKernel — APPROVED** [review]

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**